Fix class override check with contravariant Self

samwgoldman opened the pull request:
Summary:
When we check class consistency, we use the same technique as protocol matching: we substitute Self using the derived class.

Fixes https://github.com/facebook/pyrefly/issues/1052

Differential Revision: D81885918

stroxler reviewed (approved):
Review automatically exported from Phabricator review in Meta.

samwgoldman committed:
Fix class override check with contravariant Self (#1053)

Summary:
Pull Request resolved: https://github.com/facebook/pyrefly/pull/1053

When we check class consistency, we use the same technique as protocol matching: we substitute Self using the derived class.

https://github.com/facebook/pyrefly/issues/1052

Reviewed By: stroxler

Differential Revision: D81885918
